rfd/0010-api.md

```diff
@@ -0,0 +1,235 @@
+---
+authors: Alexander Klizhentas (sasha@gravitational.com), Brian Joerger (bjoerger@gravitational.com)
+state: discussion
+---
+
+# RFD 10 - API and Client Libraries
+
+## What
+
+- Document and publish API OpenAPI spec and documentation.
+- Unify Golang Client experience.
+
+## Why
+
+There are several problems with Go client libraries:
+
+* Go library example pulls entire teleport dependency including `lib/auth` and `lib/services`.
+* There is no separate client library with guarantees of compatibility with specific version of Teleport.
+See examples [here](https://github.com/kubernetes/client-go#compatibility-matrix).
+*  It's impossible to generate library in any other language other than Go due to unclear API surface.
+*  Some client logic is residing in [plugins](https://github.com/gravitational/teleport-plugins/blob/master/access/access.go), while other is in `lib/auth`.
+* Code in `lib/auth` uses some concepts unfamiliar to Go users:
+
+```go
+// connectClient establishes a gRPC connection to an auth server.
+func connectClient() (*auth.Client, error) {
+    tlsConfig, err := LoadTLSConfig("certs/api-admin.crt", "certs/api-admin.key", "certs/api-admin.cas")
+    if err != nil {
+        return nil, fmt.Errorf("Failed to setup TLS config: %v", err)
+    }
+
+    // must prase addr list panics and is not necessary
+    authServerAddr := utils.MustParseAddrList("127.0.0.1:3025")
+    clientConfig := auth.ClientConfig{Addrs: authServerAddr, TLS: tlsConfig}
+
+    // TLS client is the only client
+    return auth.NewTLSClient(clientConfig)
+```
+
+## Details
+
+### Approach
+
+Keep Teleport's API layer unified, avoid having internal API vs external API layer.
+The access to the API surface is limited by role based access control.
+
+Avoid large refactors, instead use Go's type aliases to move things around while keeping things backwards compatible.
+
+```go
+// Preserve full backwards compatibility
+type services.Role = types.Role
+type auth.Service = proto.Service
+```
+
+Use Swagger and OpenAPI to document and create SDKs.
```

Comment:
If i'm reading their docs correctly, Swagger/OpenAPI are spec and codegen tools for HTTP APIs.
They seem to be an alternative to gRPC (and not a layer on top).
Can we standardize on gRPC instead?

Reply:
Yeah it looks like OpenAPI isn't just a documentation tool, it's whole seperate protocol to express an RPC api. I'm looking into other tools to easily document our gRPC endpoints in a navigable + collapsible UI. Haven't found anything that extensive (other than google cloud endpoints which is overboard and doesn't fit well), but https://github.com/pseudomuto/protoc-gen-doc seems like a good option for now.

Reply:
protoc-gen-doc sounds good to me, especially because we can use it to autogenerate markdown files and integrate into our doc engine